sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        traces_in_sinks += len(_sev_sinks)         # Add medium- and high-severity sinks (i.e. stateID != -1)
 
-def find_severe_states(traces, m, m2):
-    med_states = set()
-    sev_states = set()
-    sev_sinks = set()
-    for i,sample in enumerate(traces):    
-        r, s, sevsinks = traverse(m, m2, sample, statelist=True)
-        sev_sinks.update(sevsinks)
-        s = s[1:]
+        assert (len(sample.split(' ')) + 1 == len(state_traces[i]))
+
+        state_list = state_list[1:]
         sample = sample.split(' ')
-        #print([(x,rev_smallmapping[x[0].split('|')[0]]) for x in zip(sample, s)])
-        med = [int(state) for sym, state in zip(sample, s) if len(str(rev_smallmapping[sym.split('|')[0]])) == 2 ]
-        med_states.update(med)
-        #print(s)
-        sev = [int(state) for sym, state in zip(sample, s) if len(str(rev_smallmapping[sym.split('|')[0]])) == 3 ]
-        #print([(sym) for sym, state in zip(sample, s) if len(str(rev_smallmapping[sym.split('|')[0]])) == 3 ])
-        sev_states.update(sev)
-        #if not set(sev_states).isdisjoint(s):
-        #    print(sample)
-        #    print('--', s)
-    #print(med_states)
-    #print(sev_states)
-    #print('med-sev traces')
-    #for i,sample in enumerate(traces):
-    med_states = med_states.difference(sev_states)  
-        
-    #    r, s = traverse(m, m2, sample, statelist=True)
-    #    s = [int(x) for x in s]
-    #    #print(s)
-    #    if not set(med_states).isdisjoint(s):
-    #        print(sample)
-    #        print('--', s)
-    print('Total medium states', len(med_states))  
-    print('Total severe states', len(sev_states))
-    return(med_states, sev_states, sev_sinks)
-
-## collecting sub-behaviors back into the same trace -- condensed_data is the new object to deal with
-def make_condensed_data(alerts, keys, state_traces, med_states, sev_states):
-    levelone = set()
-    levelone_ben = set()
-    condensed_data = dict()
+        med_sev = [int(state) for sym, state in zip(sample, state_list) if len(str(rev_smallmapping[sym.split('|')[0]])) == 2]
+        med_sev_states.update(med_sev)
+        high_sev = [int(state) for sym, state in zip(sample, state_list) if len(str(rev_smallmapping[sym.split('|')[0]])) == 3]
+        high_sev_states.update(high_sev)
+        sev_sinks.update(_sev_sinks)
+
+    print('Traces in sinks:', traces_in_sinks, 'Total traces:', total_traces, 'Percentage:', 100 * (traces_in_sinks / float(total_traces)))
+    print('Total medium-severity states:', len(med_sev_states))
+    print('Total high-severity states:', len(high_sev_states))
+    print('Total severe sinks:', len(sev_sinks))
+    return state_traces, med_sev_states, high_sev_states, sev_sinks
+
+
+# Step 6.3: Create state sequences (collecting sub-behaviors back into the same trace, augmented with state IDs)
+def make_state_sequences(episode_subsequences, state_traces, med_states, sev_states):
```

Comment:
`condenced_data` is essentially state sequences, from what I can see in the paper

Reply:
Yes